File: apple/CaCompteKit/Tests/CatalogTests/GoldenFile.swift
import Domain
import Foundation
import Testing

/// Décodage de `spec/golden/*.json` — format documenté dans `spec/README.md`. Les identifiants
/// de participants ("p1", "p2"…) sont des jetons de test lisibles, pas des `UUID` : le test les
/// convertit à la volée.
struct GoldenFile: Decodable, Sendable {
  struct Participant: Decodable, Sendable {
    let id: String
    let name: String
    let seatIndex: Int
    /// Doc 05 « Belote » — `nil` pour tous les jeux individuels.
    let teamID: String?
  }

  struct Input: Decodable, Sendable {
    let participant: String
    let rawValue: Int
    let modifiers: [String]
    /// Payload de `ScoreDetail`, générique dans ce format (`spec/` ignore Catalog) — la
    /// traduction vers un type concret (ex. `YamsCategoryDetail`) se fait dans le test,
    /// pas ici.
    let detail: [String: String]?
  }

  struct RoundInput: Decodable, Sendable {
    let index: Int
    let inputs: [Input]
  }

  struct RoundResult: Decodable, Sendable {
    let index: Int
    let computed: [String: Int]
    let cumulative: [String: Int]
    let doubled: [String]
    let status: String
  }

  struct StandingResult: Decodable, Sendable {
    let participant: String
    let rank: Int
    let score: Int
    let sharedWith: [String]?
  }

  struct FinalResult: Decodable, Sendable {
    let status: String
    let reason: String?
    let standings: [StandingResult]
  }

  struct InsightExpectation: Decodable, Sendable {
    let id: String
    let participant: String?
    let value: Double?
    let round: Int?
    let values: [String: Double]?
  }

  struct Expected: Decodable, Sendable {
    let roundResults: [RoundResult]
    let final: FinalResult
    let insights: [InsightExpectation]
  }

  let goldenId: String
  let gameId: String
  let rulesVersion: Int
  let variants: VariantSelection
  let participants: [Participant]
  let rounds: [RoundInput]
  let expected: Expected

  static let all: [GoldenFile] = {
    let urls =
      (Bundle.module.urls(forResourcesWithExtension: "json", subdirectory: "GoldenResources") ?? [])
      .sorted { $0.lastPathComponent < $1.lastPathComponent }
    let decoder = JSONDecoder()
    return urls.map { url in
      let data = try! Data(contentsOf: url)
      return try! decoder.decode(GoldenFile.self, from: data)
    }
  }()
}

extension GoldenFile: CustomTestStringConvertible {
  var testDescription: String { goldenId }
}

File: android/catalog/src/test/kotlin/com/quimene/catalog/golden/GoldenFile.kt
package com.quimene.catalog.golden

import com.quimene.domain.model.VariantSelection
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import java.io.File

/**
 * Miroir de `GoldenFile.swift` — décodage des fichiers `.json` de `spec/golden` (format documenté dans
 * `spec/README.md`). Les identifiants de participants ("p1", "p2"…) sont des jetons de test
 * lisibles, pas des `UUID` : [com.quimene.catalog.golden.GoldenFileTest] les convertit à la
 * volée. `ignoreUnknownKeys` : les golden files portent un champ `description` documentaire non
 * modélisé ici, comme le décodage Swift par défaut l'ignore silencieusement.
 */
@Serializable
data class GoldenFile(
    val goldenId: String,
    val gameId: String,
    val rulesVersion: Int,
    val variants: VariantSelection,
    val participants: List<Participant>,
    val rounds: List<RoundInput>,
    val expected: Expected,
) {
    @Serializable
    data class Participant(
        val id: String,
        val name: String,
        val seatIndex: Int,
        val teamID: String? = null,
    )

    @Serializable
    data class Input(
        val participant: String,
        val rawValue: Int,
        val modifiers: List<String> = emptyList(),
        val detail: Map<String, String>? = null,
    )

    @Serializable
    data class RoundInput(
        val index: Int,
        val inputs: List<Input>,
    )

    @Serializable
    data class RoundResult(
        val index: Int,
        val computed: Map<String, Int>,
        val cumulative: Map<String, Int>,
        val doubled: List<String> = emptyList(),
        val status: String,
    )

    @Serializable
    data class StandingResult(
        val participant: String,
        val rank: Int,
        val score: Int,
        val sharedWith: List<String>? = null,
    )

    @Serializable
    data class FinalResult(
        val status: String,
        val reason: String? = null,
        val standings: List<StandingResult>,
    )

    @Serializable
    data class InsightExpectation(
        val id: String,
        val participant: String? = null,
        val value: Double? = null,
        val round: Int? = null,
        val values: Map<String, Double>? = null,
    )

    @Serializable
    data class Expected(
        val roundResults: List<RoundResult>,
        val final: FinalResult,
        val insights: List<InsightExpectation>,
    )

    override fun toString(): String = goldenId

    companion object {
        private val json = Json { ignoreUnknownKeys = true }

        /** Miroir de `GoldenFile.all` — balaie la ressource `GoldenResources` régénérée par la
         * tâche Gradle `copyGoldenResources`, triée par nom de fichier. */
        val all: List<GoldenFile> by lazy {
            val resourceUrl =
                requireNotNull(GoldenFile::class.java.classLoader.getResource("GoldenResources")) {
                    "Ressource GoldenResources introuvable — copyGoldenResources a-t-il tourné ?"
                }
            val directory = File(resourceUrl.toURI())
            val files =
                directory.listFiles { file -> file.extension == "json" }
                    ?: error("GoldenResources n'est pas un dossier lisible : $directory")

            files.sortedBy { it.name }.map { file -> json.decodeFromString<GoldenFile>(file.readText()) }
        }
    }
}
